src/Utilities/variables.jl

```diff
@@ -13,45 +13,107 @@ Return a tuple `(lb, ub)` of type `Tuple{T, T}`, where `lb` and `ub` are lower
+    if MOI.is_valid(model, c_interval)
+        s_interval = MOI.get(model, MOI.ConstraintSet(), c_interval)
+        return s_interval.lower, s_interval.upper
+    end
+    # MOI.EqualTo
+    c_equal_to = MOI.ConstraintIndex{F,MOI.EqualTo{T}}(x.value)
+    if MOI.is_valid(model, c_equal_to)
+        s_equal_to = MOI.get(model, MOI.ConstraintSet(), c_equal_to)
+        return s_equal_to.value, s_equal_to.value
+    end
+    # MOI.Semicontinuous
+    c_semicontinuous = MOI.ConstraintIndex{F,MOI.Semicontinuous{T}}(x.value)
+    if MOI.is_valid(model, c_semicontinuous)
+        s_semicontinuous = MOI.get(model, MOI.ConstraintSet(), c_semicontinuous)
+        l = min(zero(T), s_semicontinuous.lower)
+        u = max(zero(T), s_semicontinuous.upper)
+        return l, u
+    end
+    # MOI.Semiinteger
+    c_si = MOI.ConstraintIndex{F,MOI.Semiinteger{T}}(x.value)
+    if MOI.is_valid(model, c_si)
         si::MOI.Semiinteger{T} = MOI.get(model, MOI.ConstraintSet(), c_si)
         return min(zero(T), si.lower), max(zero(T), si.upper)
-    elseif MOI.is_valid(model, c_lt)
-        lt::MOI.LessThan{T} = MOI.get(model, MOI.ConstraintSet(), c_lt)
-        # It is valid to have both LessThan and GreaterThan constraints on the
-        # same variable.
-        if MOI.is_valid(model, c_gt)
-            gt_1::MOI.GreaterThan{T} = MOI.get(model, MOI.ConstraintSet(), c_gt)
-            return gt_1.lower, lt.upper
-        else
-            return typemin(T), lt.upper
+    end
+    l, u = typemin(T), typemax(T)
+    # MOI.LessThan
+    c_less_than = MOI.ConstraintIndex{F,MOI.LessThan{T}}(x.value)
+    if MOI.is_valid(model, c_less_than)
+        s_less_than = MOI.get(model, MOI.ConstraintSet(), c_less_than)
+        u = min(u, s_less_than.upper)
+    end
+    # MOI.GreaterThan
+    c_greater_than = MOI.ConstraintIndex{F,MOI.GreaterThan{T}}(x.value)
+    if MOI.is_valid(model, c_greater_than)
+        s_greater_than = MOI.get(model, MOI.ConstraintSet(), c_greater_than)
+        l = max(l, s_greater_than.lower)
+    end
+    return l, u
+end
+
+"""
+    get_bounds(
+        model::MOI.ModelLike,
+        bounds_cache::Dict{MOI.VariableIndex,NTuple{2,T}},
+        f::MOI.ScalarAffineFunction{T},
+    ) where {T} --> Union{Nothing,NTuple{2,T}}
+
+Return the lower and upper bound of `f` as a tuple. If the domain is not bounded,
+return `nothing`.
+"""
+function get_bounds(
+    model::MOI.ModelLike,
+    bounds_cache::Dict{MOI.VariableIndex,NTuple{2,T}},
+    f::MOI.ScalarAffineFunction{T},
+) where {T}
+    lb = ub = f.constant
+    for term in f.terms
+        ret = get_bounds(model, bounds_cache, term.variable)
+        if ret === nothing
+            return nothing
         end
-    elseif MOI.is_valid(model, c_gt)
-        gt_2::MOI.GreaterThan{T} = MOI.get(model, MOI.ConstraintSet(), c_gt)
-        return gt_2.lower, typemax(T)
-    else
-        return typemin(T), typemax(T)
+        lb += term.coefficient * ret[1]
```

Comment:
If `x in [-1, 2]` then you will say that `-x` is in `[1, -2]`. If the sign of `term.coefficient` is negative, you should reverse the bounds

Reply:
Again if you have `x - 2x` then you'll get as lb `-1 - 2 = -3` and ub `2 - 2 * (-1) = 4` while if you canonicalize first you have `-2, 1` which is tighter. So I would canonicalize first.

Reply:
Hmm. I obviously don't have tests for this

Reply:
By the way, PolyJuMP.QCQP would be a user of a `get_bounds` for `ScalarQuadraticTerm`: https://github.com/jump-dev/PolyJuMP.jl/blob/aee87cda22f76545b0a611e87cb753cbc1d786f0/src/QCQP/MOI_wrapper.jl#L264-L271

Reply:
One argument for centralizing the implementation, like I've done in this PR

Reply:
Yes, we reached the threshold that it's useful to centralize

Reply:
Fixed